docs/developer/api-versioning.md

```diff
@@ -0,0 +1,391 @@
+too old" exception, and gracefully shut down the new functionality.
+
+
+## Concerns and design alternatives
+
+### Why not version every end-point separately?
+
+Yes, that would work in principle.  On the backend, it would make the
+entire routing table smaller (no need to concatenate the same
+end-point for many versions), which may result in shorter compile
+times.  On the clients, with a new API version it would be
+straight-forward to see which end-points need to be worked on, and
+which remain unchanged.
+
+On the other hand, the routing table size may not be an issue, and if
+it is there are solutions (introduce a CCP switch to compile only the
+most recent API version that you're working on); and the client
+process is already quite straight-forward with the approach outlined
+above via diffing the swagger docs between most recent version and
+predecessor.
+
+Plus, if the entire API has one version, you get a few advantages:
+
+1. The fact that clients are forced to commit to a concrete API
+   version for all end-points when talking to the backend reduces
+   testing complexity.  If there is a mapping of end-points to
+   versions, the behavior of interacting parties is much less
+   restricted, and versions that have not been tested against each
+   other may be used together.  (This can be avoided, but it's less
+   obvious how to get it right, and testing complexity will likely be
+   worse.)
+
+2. The "one version" approach makes it obvious which end-points are in
+   the most recent API at any given point in time.  The "one version
+   per end-point" approach would either yeild a noisy union of all
+   supported versions, or there would have to be a mechanism for
+   reconstructing something close to what we get for free otherwise.
+
+3. The backend code is a good combination of concise and type-safe in
+   the "one version" approach.  If every end-point had its own
+   version, the routing table entry would either have to accept a
+   variable path segment for the version, and fail at run-time if the
+   version is not supported, or you would have to add one handler per
+   supported version (even if in the case where all versions call the
+   same handler function with slightly different parameters).
+
+
+### Syntactical vs. behavioral changes
+
+It is quite common that behavior of end-points changes together with
+the syntax, or even without a change in the syntax.
+
+This is not a fundamental problem: since the handler can be called
+with the version as a type parameter, there is no reason why it
+shouldn't change behavior with or without changing the syntax.  In
+each such case, it needs to be decided whether the difference is
+significant enough to justify a new API version.
+
+At the very least though it should result in diverging swagger docs
+that explains those differences.
+
+
+### Client capabilities
+
+Wire supports client capabilities to decide whether a client should be
+allowed to use certain parts of the API.
```

Comment:
Can you elaborate on what you mean by "client capabilities"? Is it feature flags? You mean those compiled in clients (e.g. for customer-specific builds), or backend-controlled feature flags, such as for teams that have some features turned on/off?

Reply:
backend-controlled feature flags

Reply:
Nope, it's another thing that we introduce for LegalHold.  we want clients that have no LegalHold capability to be blocked from every getting tainted by a LH device (or a user that has one).  So we introduced capabilities in the client schema, and clients announce them: https://staging-nginz-https.zinfra.io/api/swagger-ui/#/default/get_clients

Anyway, this is something we want to eliminate, but don't need to, so it's not central to the debate.